test/test_pytree.py

```diff
@@ -60,35 +42,55 @@ def run_test_with_leaf(leaf):
         run_test_with_leaf(bool)
         run_test_with_leaf(torch.randn(3, 3))
 
-    def test_flatten_unflatten_list(self):
+    @parametrize(
+        "pytree_impl,gen_expected_fn",
+        [
+            (
+                py_pytree,
+                lambda lst: py_pytree.TreeSpec(
+                    list, None, [py_pytree.LeafSpec() for _ in lst]
+                ),
+            ),
+            (cxx_pytree, lambda lst: cxx_pytree.tree_structure([0] * len(lst))),
+        ],
+    )
+    def test_flatten_unflatten_list(self, pytree_impl, gen_expected_fn):
```

Comment:
Why do we build the TreeSpec for Python but instead use tree_structure to construct it for cxx_pytree? Does cxx_pytree not have a way to manually build a TreeSpec?

Reply:
> Does cxx_pytree not have a way to manually build a TreeSpec?

In `optree`, no. The only ways to build an arbitrary TreeSpec are the `tree_flatten` or `tree_flatten_with_path` functions. There is no way to build a TreeSpec from the Python side. We test it via the roundtrip property in `optree`.

Note: `optree` do support creating leaf, `None`, and `tuple` types from children treespecs. But other types like `list`, `dict`, `OrderedDict`, etc, are not implemented yet.

Reply:
TreeSpec is technically a part of the Python pytree API. So I'm not really sure what we're going to do with it. Thoughts?

Reply:
`optree` does not provide the Python side APIs (at least not all types) to create a PyTreeSpec object. Because the user-provided node `metadata` for custom types may be malformed. Also, there may be other potential bugs that the user-provided `keys` for dict nodes are not hashable or not unique. If only for test propose and all APIs to a build TreeSpec are private, I think it's OK to add them.

Reply:
If we want to make both python and C++ pytree public, then either they should be different APIs or we should unify TreeSpec. (For example, if our strategy is that C++ pytree falls back to Python pytree). TreeSpec is technically public API, it would be problematic if someone does an `isinstance` check.

We should figure out how to deal with this when the time comes (so, not in this PR).

Reply:
> I agree that the access to TreeSpec type should be public. But I prefer to make the attributes of TreeSpec private and only public some methods to the users. All attributes of TreeSpec should be private and immutable (we did this in C++ pytree). If we don't, the user can manipulate the treespec attributes and make them malformed.

This will be mildly bc-breaking, but I agree in general